# Домашнее задание к лекции "Базовые понятия статистики"

## Задание 1. Базовое изучение
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [4]:
import pandas as pd

In [2]:
import numpy as np

In [5]:
colnames=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28] 
horse_colic_database = pd.read_csv('horse_data.csv', names = colnames, header=None, na_values='?')
horse_colic_database

1   2        3     4      5     6    7    8    9    10  ...    19    20  \
0    2.0   1   530101  38.5   66.0  28.0  3.0  3.0  NaN  2.0  ...  45.0   8.4   
1    1.0   1   534817  39.2   88.0  20.0  NaN  NaN  4.0  1.0  ...  50.0  85.0   
2    2.0   1   530334  38.3   40.0  24.0  1.0  1.0  3.0  1.0  ...  33.0   6.7   
3    1.0   9  5290409  39.1  164.0  84.0  4.0  1.0  6.0  2.0  ...  48.0   7.2   
4    2.0   1   530255  37.3  104.0  35.0  NaN  NaN  6.0  2.0  ...  74.0   7.4   
..   ...  ..      ...   ...    ...   ...  ...  ...  ...  ...  ...   ...   ...   
295  1.0   1   533886   NaN  120.0  70.0  4.0  NaN  4.0  2.0  ...  55.0  65.0   
296  2.0   1   527702  37.2   72.0  24.0  3.0  2.0  4.0  2.0  ...  44.0   NaN   
297  1.0   1   529386  37.5   72.0  30.0  4.0  3.0  4.0  1.0  ...  60.0   6.8   
298  1.0   1   530612  36.5  100.0  24.0  3.0  3.0  3.0  1.0  ...  50.0   6.0   
299  1.0   1   534618  37.2   40.0  20.0  NaN  NaN  NaN  NaN  ...  36.0  62.0   

      21   22   23  24     25  26  27  28  
0    NaN  NaN  2.0   2  11300   0   0   2  
1    2.0  2.0  3.0   2   2208   0   0   2  
2    NaN  NaN  1.0   2      0   0   0   1  
3    3.0  5.3  2.0   1   2208   0   0   1  
4    NaN  NaN  2.0   2   4300   0   0   2  
..   ...  ...  ...  ..    ...  ..  ..  ..  
295  NaN  NaN  3.0   2   3205   0   0   2  
296  3.0  3.3  3.0   1   2208   0   0   1  
297  NaN  NaN  2.0   1   3205   0   0   2  
298  3.0  3.4  1.0   1   2208   0   0   1  
299  1.0  1.0  3.0   2   6112   0   0   2  

[300 rows x 28 columns]

In [7]:
my_horse_colic_database = horse_colic_database.iloc[:, [0,1,3,4,5,6,7,8]]
my_horse_colic_database

1  2     4      5     6    7    8    9
0    2.0  1  38.5   66.0  28.0  3.0  3.0  NaN
1    1.0  1  39.2   88.0  20.0  NaN  NaN  4.0
2    2.0  1  38.3   40.0  24.0  1.0  1.0  3.0
3    1.0  9  39.1  164.0  84.0  4.0  1.0  6.0
4    2.0  1  37.3  104.0  35.0  NaN  NaN  6.0
..   ... ..   ...    ...   ...  ...  ...  ...
295  1.0  1   NaN  120.0  70.0  4.0  NaN  4.0
296  2.0  1  37.2   72.0  24.0  3.0  2.0  4.0
297  1.0  1  37.5   72.0  30.0  4.0  3.0  4.0
298  1.0  1  36.5  100.0  24.0  3.0  3.0  3.0
299  1.0  1  37.2   40.0  20.0  NaN  NaN  NaN

[300 rows x 8 columns]

In [73]:
my_horse_colic_database_2 = my_horse_colic_database.set_axis([
    '_1_Surgery', 
    '_2_Age',
    '_4_Temperature',
    '_5_Pulse',
    '_6_Respiratory_rate',
    '_7_Temperature_of_extremities',
    '_8_Peripheral_pulse',
    '_9_Mucous_membranes',
], axis=1, inplace=False)
my_horse_colic_database_2

_1_Surgery  _2_Age  _4_Temperature  _5_Pulse  _6_Respiratory_rate  \
0           2.0       1            38.5      66.0                 28.0   
1           1.0       1            39.2      88.0                 20.0   
2           2.0       1            38.3      40.0                 24.0   
3           1.0       9            39.1     164.0                 84.0   
4           2.0       1            37.3     104.0                 35.0   
..          ...     ...             ...       ...                  ...   
295         1.0       1             NaN     120.0                 70.0   
296         2.0       1            37.2      72.0                 24.0   
297         1.0       1            37.5      72.0                 30.0   
298         1.0       1            36.5     100.0                 24.0   
299         1.0       1            37.2      40.0                 20.0   

     _7_Temperature_of_extremities  _8_Peripheral_pulse  _9_Mucous_membranes  
0                              3.0                  3.0                  NaN  
1                              NaN                  NaN                  4.0  
2                              1.0                  1.0                  3.0  
3                              4.0                  1.0                  6.0  
4                              NaN                  NaN                  6.0  
..                             ...                  ...                  ...  
295                            4.0                  NaN                  4.0  
296                            3.0                  2.0                  4.0  
297                            4.0                  3.0                  4.0  
298                            3.0                  3.0                  3.0  
299                            NaN                  NaN                  NaN  

[300 rows x 8 columns]

In [74]:
# Согласно описанию столбец 2 может принимать значения 1 и 2, где 1 - взрослая лошадь, 2 - молодая лошадь.
# В наших данных в этом столбце предсталены значение 1 и 9. Вероятно произошла ошибка и значение 9 это 2.
# Заменим значение 9 на 2


my_horse_colic_database_2['_2_Age'].replace(9, 2, inplace=True)
my_horse_colic_database_2

_1_Surgery  _2_Age  _4_Temperature  _5_Pulse  _6_Respiratory_rate  \
0           2.0       1            38.5      66.0                 28.0   
1           1.0       1            39.2      88.0                 20.0   
2           2.0       1            38.3      40.0                 24.0   
3           1.0       2            39.1     164.0                 84.0   
4           2.0       1            37.3     104.0                 35.0   
..          ...     ...             ...       ...                  ...   
295         1.0       1             NaN     120.0                 70.0   
296         2.0       1            37.2      72.0                 24.0   
297         1.0       1            37.5      72.0                 30.0   
298         1.0       1            36.5     100.0                 24.0   
299         1.0       1            37.2      40.0                 20.0   

     _7_Temperature_of_extremities  _8_Peripheral_pulse  _9_Mucous_membranes  
0                              3.0                  3.0                  NaN  
1                              NaN                  NaN                  4.0  
2                              1.0                  1.0                  3.0  
3                              4.0                  1.0                  6.0  
4                              NaN                  NaN                  6.0  
..                             ...                  ...                  ...  
295                            4.0                  NaN                  4.0  
296                            3.0                  2.0                  4.0  
297                            4.0                  3.0                  4.0  
298                            3.0                  3.0                  3.0  
299                            NaN                  NaN                  NaN  

[300 rows x 8 columns]

## Attribute Information:
1:  surgery?
          1 = Yes, it had surgery
          2 = It was treated without surgery

  2:  Age
          1 = Adult horse
          2 = Young (< 6 months)

  3:  Hospital Number
          - numeric id
          - the case number assigned to the horse
            (may not be unique if the horse is treated > 1 time)

  4:  rectal temperature
          - linear
          - in degrees celsius.
          - An elevated temp may occur due to infection.
          - temperature may be reduced when the animal is in late shock
          - normal temp is 37.8
          - this parameter will usually change as the problem progresses
               eg. may start out normal, then become elevated because of
                   the lesion, passing back through the normal range as the
                   horse goes into shock
  5:  pulse
          - linear
          - the heart rate in beats per minute
          - is a reflection of the heart condition: 30 -40 is normal for adults
          - rare to have a lower than normal rate although athletic horses
            may have a rate of 20-25
          - animals with painful lesions or suffering from circulatory shock
            may have an elevated heart rate

  6:  respiratory rate
          - linear
          - normal rate is 8 to 10
          - usefulness is doubtful due to the great fluctuations

  7:  temperature of extremities
          - a subjective indication of peripheral circulation
          - possible values:
               1 = Normal
               2 = Warm
               3 = Cool
               4 = Cold
          - cool to cold extremities indicate possible shock
          - hot extremities should correlate with an elevated rectal temp.

  8:  peripheral pulse
          - subjective
          - possible values are:
               1 = normal
               2 = increased
               3 = reduced
               4 = absent
          - normal or increased p.p. are indicative of adequate circulation
            while reduced or absent indicate poor perfusion

  9:  mucous membranes
          - a subjective measurement of colour
          - possible values are:
               1 = normal pink
               2 = bright pink
               3 = pale pink
               4 = pale cyanotic
               5 = bright red / injected
               6 = dark cyanotic
          - 1 and 2 probably indicate a normal or slightly increased
            circulation
          - 3 may occur in early shock
          - 4 and 6 are indicative of serious circulatory compromise
          - 5 is more indicative of a septicemia

In [75]:
my_horse_colic_database_2.describe()

_1_Surgery      _2_Age  _4_Temperature    _5_Pulse  \
count  299.000000  300.000000      240.000000  276.000000   
mean     1.397993    1.080000       38.167917   71.913043   
std      0.490305    0.271746        0.732289   28.630557   
min      1.000000    1.000000       35.400000   30.000000   
25%      1.000000    1.000000       37.800000   48.000000   
50%      1.000000    1.000000       38.200000   64.000000   
75%      2.000000    1.000000       38.500000   88.000000   
max      2.000000    2.000000       40.800000  184.000000   

       _6_Respiratory_rate  _7_Temperature_of_extremities  \
count           242.000000                     244.000000   
mean             30.417355                       2.348361   
std              17.642231                       1.045054   
min               8.000000                       1.000000   
25%              18.500000                       1.000000   
50%              24.500000                       3.000000   
75%              36.000000                       3.000000   
max              96.000000                       4.000000   

       _8_Peripheral_pulse  _9_Mucous_membranes  
count           231.000000           253.000000  
mean              2.017316             2.853755  
std               1.042428             1.620294  
min               1.000000             1.000000  
25%               1.000000             1.000000  
50%               2.000000             3.000000  
75%               3.000000             4.000000  
max               4.000000             6.000000

In [76]:
print('Значение моды Surgery:', my_horse_colic_database_2['_1_Surgery'].round().mode()[0])
print('Значение моды Age :', my_horse_colic_database_2['_2_Age'].round().mode()[0])
print('Значение моды Temperature:', my_horse_colic_database_2['_4_Temperature'].round(decimals=1).mode()[0])
print('Значение моды Pulse:', my_horse_colic_database_2['_5_Pulse'].round().mode()[0])
print('Значение моды Respiratory rate:', my_horse_colic_database_2['_6_Respiratory_rate'].round().mode()[0])
print('Значение моды Temperature of extremities:', my_horse_colic_database_2['_7_Temperature_of_extremities'].round().mode()[0])
print('Значение моды Peripheral pulse:', my_horse_colic_database_2['_8_Peripheral_pulse'].round().mode()[0])
print('Значение моды Mucous membranes:', my_horse_colic_database_2['_9_Mucous_membranes'].round().mode()[0])


Значение моды Surgery: 1.0
Значение моды Age : 1
Значение моды Temperature: 38.0
Значение моды Pulse: 48.0
Значение моды Respiratory rate: 20.0
Значение моды Temperature of extremities: 3.0
Значение моды Peripheral pulse: 1.0
Значение моды Mucous membranes: 1.0


Краткое описание результатов:
- значения столбцов 1, 2, 7, 8, 9 являются категориальными
- значения столбца 4 являются непрерывными
- значения столбцов 5, 6 являются дискретными
Общее колчиество лошадей (строк) - 300
Остальные найденные параметры (кроме моды) для категорийных значений смысла в себе не несут
Температура (4):
- среднее значение 38,17
- СКО (среднее квадратичное отклонение) 0,73
- минимальное значение 35,4
- 0,25 кванитль 37,8
- 0,5 квантиль (медиана) 38,2
- 0,75 квантиль 38,5
- максимальное значение 40,8
Аналогично для 5 и 6


- (1) Surgery: большая часть лошадей обошлась без операции (значение моды для столбц 1 равно 1)
- (2) Age: большая часть лошдей взрослые 
- (4) Temperature: самая часто встречаемая -ра 38, при нормальной 37,8. И средняя т-ра 38,18. Веротяно большая часть поголовья имеет инфекцию, которая вызвала повышение тем-ры
- (5) Pulse: мода пульса 48 (нормальное значение 30-40), среднее значение 71,9. Но мы вилим анамально высокое максимальное значение 184, что может говорить о выбросах, которые повлияли на результаты 
- (6) Respiratory rate: мода частоты дыхания 20 и срднее значение 30,4 (при номальном от 8 до 10). В описании написано, что данных парментр имеет сомнительную важность, но повышенные значения косвенно подтверждают нездоровое состояние большинства животных
- (7) Temperature of extremities: мода температуры конечностей 3 (cool, прохладные), что указывает на возможное шоковое состояние большинсвта животных
- (8) Peripheral pulse:самое частое значение периферического пульса 1 (нормальное)
- (9) mucous membranes: самое частая оценка слизистой оболочки 1, т.е. нормальное розовое

В целом первчиная оцнека показателей говорит о вероятном болезненном состоянии большей части поголовья лошадей.


## Задание 2. Работа с выбросами
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

Значения столбцов 1, 2, 7, 8, 9 являются категориальными - рассчитывать для них выбросы не имеет смысла. Проверим оишь что бы все значение были среди категорий для данного параметра (все значения соответствуют заявоенным категориям.

Найдем выбросы для столбцов [4, Temperature], 	[5, Pulse],	[6, Respiratory rate]

#### Рассмотрим столбец [4, Temperature]
Нормальное значение температуры (здоровое животное) - 37,8
Минимальное значение - 35,4
Максимальное - 40,8

Экстримальные значение мин и макс находятся в разумных пределах. Считаем, что выбрасов в данном столбце нет.

#### Рассмотрим столбец [5, Pulse]
Нормальное значение 30-40 ударов.
У молодых особей нормальный пульс 70 - 120 ударов в минуту.
Минимальное значение 30 - не вызывает подозрений, очевидно не выброс.
Максимальное значение 184 - тоже находится в разумных пределах, т.е. это не нереальный показатель. Веротяно повышенный пульс говорит о наличии заболевания. 
Я считаю, что в данном столбце тоже нет очевыидных выбросов. Но тем не менее найдем выбросы через межквартльный размах и проанализируем их:


In [77]:
q1_5 = my_horse_colic_database_2['_5_Pulse'].quantile(0.25)
q3_5 = my_horse_colic_database_2['_5_Pulse'].quantile(0.75)
iqr_5 = q3_5 - q1_5
lower_bound_5 = q1_5 - (1.5 * iqr_5) 
upper_bound_5 = q3_5 + (1.5 * iqr_5)
print(q1_5, q3_5, iqr_5, lower_bound_5, upper_bound_5)
remove_outliers = my_horse_colic_database_2[my_horse_colic_database_2['_5_Pulse'].between(lower_bound_5, upper_bound_5, inclusive=True)]
# remove_outliers
pd.concat([my_horse_colic_database_2, remove_outliers]).drop_duplicates(keep=False)

48.0 88.0 40.0 -12.0 148.0


_1_Surgery  _2_Age  _4_Temperature  _5_Pulse  _6_Respiratory_rate  \
3           1.0       2            39.1     164.0                 84.0   
5           2.0       1             NaN       NaN                  NaN   
41          2.0       2            39.0     150.0                 72.0   
52          2.0       1             NaN       NaN                  NaN   
55          1.0       2            38.6     160.0                 20.0   
58          1.0       1             NaN       NaN                 20.0   
93          2.0       1             NaN       NaN                  NaN   
115         2.0       1             NaN       NaN                 40.0   
117         2.0       1            39.5       NaN                  NaN   
126         1.0       1            38.7       NaN                  NaN   
151         1.0       1            39.3       NaN                  NaN   
159         2.0       1             NaN       NaN                 12.0   
174         2.0       1             NaN       NaN                  NaN   
204         1.0       1            39.2       NaN                 23.0   
216         2.0       1             NaN       NaN                  NaN   
255         1.0       2            38.8     184.0                 84.0   
275         1.0       2            38.8     150.0                 50.0   

     _7_Temperature_of_extremities  _8_Peripheral_pulse  _9_Mucous_membranes  
3                              4.0                  1.0                  6.0  
5                              2.0                  1.0                  3.0  
41                             NaN                  NaN                  NaN  
52                             1.0                  1.0                  3.0  
55                             3.0                  NaN                  5.0  
58                             4.0                  3.0                  3.0  
93                             3.0                  3.0                  1.0  
115                            3.0                  1.0                  1.0  
117                            3.0                  3.0                  4.0  
126                            3.0                  1.0                  3.0  
151                            4.0                  3.0                  6.0  
159                            1.0                  1.0                  2.0  
174                            3.0                  4.0                  6.0  
204                            3.0                  1.0                  3.0  
216                            NaN                  NaN                  NaN  
255                            1.0                  NaN                  1.0  
275                            1.0                  3.0                  6.0

По результатам мы видим, что повышеные значения пульса встречаются только у молодых особей. У жеребят нормальный пульс 70 - 120 ударов в минуту. Веротяно повышенная частота пульса говорит о болезни этих животных или о возможных физ нагрузказ перед измерением. Пологаю, что нет основыний считать данные значения выбросами.

#### Рассмотрим столбец [6, Respiratory rate] (частота дыхания)

Согласно описанию к данным:
Частота дыхания
- линейный
- нормальная оценка от 8 до 10
- полезность сомнительна из-за больших колебаний

Минимальное значение данного параметра 8 - находится нормальном диапозоне.
Максимальное значение 96 - вероятнее всего является выбросом.

Произведем рассчеты и проанализируем результаты:

In [78]:
q1_6 = my_horse_colic_database_2['_6_Respiratory_rate'].quantile(0.25)
q3_6 = my_horse_colic_database_2['_6_Respiratory_rate'].quantile(0.75)
iqr_6 = q3_6 - q1_6
lower_bound_6 = q1_6 - (1.5 * iqr_6) 
upper_bound_6 = q3_6 + (1.5 * iqr_6)
print(q1_6, q3_6, iqr_6, lower_bound_6, upper_bound_6)
remove_outliers_6 = my_horse_colic_database_2[my_horse_colic_database_2['_6_Respiratory_rate'].between(lower_bound_6, upper_bound_6, inclusive=True)]
# remove_outliers
pd.concat([my_horse_colic_database_2, remove_outliers_6]).drop_duplicates(keep=False)

18.5 36.0 17.5 -7.75 62.25


_1_Surgery  _2_Age  _4_Temperature  _5_Pulse  _6_Respiratory_rate  \
3           1.0       2            39.1     164.0                 84.0   
5           2.0       1             NaN       NaN                  NaN   
7           1.0       1             NaN      60.0                  NaN   
9           2.0       2            38.3      90.0                  NaN   
29          2.0       1            37.7      48.0                  NaN   
32          1.0       1            37.2      60.0                  NaN   
37          1.0       1            37.8      72.0                  NaN   
38          2.0       1            38.6      52.0                  NaN   
39          1.0       2            39.2     146.0                 96.0   
40          1.0       1             NaN      88.0                  NaN   
41          2.0       2            39.0     150.0                 72.0   
43          1.0       1             NaN     120.0                  NaN   
45          2.0       1             NaN     120.0                  NaN   
50          2.0       1            38.6      46.0                  NaN   
52          2.0       1             NaN       NaN                  NaN   
61          2.0       1            38.5      60.0                  NaN   
66          2.0       1            37.9      72.0                  NaN   
72          1.0       1            37.7      56.0                  NaN   
75          1.0       2            39.7     100.0                  NaN   
81          1.0       1            37.3      40.0                  NaN   
82          1.0       2            38.1     100.0                 80.0   
84          1.0       1            37.8      60.0                 80.0   
93          2.0       1             NaN       NaN                  NaN   
103         1.0       2            38.0     140.0                 68.0   
106         1.0       1            38.3      52.0                 96.0   
112         1.0       1             NaN      40.0                  NaN   
117         2.0       1            39.5       NaN                  NaN   
120         1.0       1            39.4      54.0                 66.0   
125         1.0       1            38.0      42.0                 68.0   
126         1.0       1            38.7       NaN                  NaN   
129         1.0       1            38.4      49.0                  NaN   
134         1.0       1             NaN     100.0                  NaN   
137         1.0       1            38.0      48.0                  NaN   
138         2.0       1            38.0      56.0                  NaN   
142         1.0       1             NaN     120.0                  NaN   
146         1.0       1            38.2      48.0                  NaN   
148         1.0       1            38.3     132.0                  NaN   
151         1.0       1            39.3       NaN                  NaN   
152         1.0       1             NaN     100.0                  NaN   
164         1.0       1            38.7      76.0                  NaN   
173         1.0       1             NaN     104.0                  NaN   
174         2.0       1             NaN       NaN                  NaN   
188         1.0       1            38.5      60.0                  NaN   
201         1.0       1            38.5      86.0                  NaN   
208         1.0       1            37.8      88.0                 80.0   
216         2.0       1             NaN       NaN                  NaN   
229         1.0       2            38.5     120.0                 70.0   
238         2.0       1            36.1      88.0                  NaN   
244         1.0       2            38.2     124.0                 88.0   
249         2.0       1            38.2      48.0                  NaN   
254         1.0       1             NaN      80.0                  NaN   
255         1.0       2            38.8     184.0                 84.0   
256         1.0       1            37.5      72.0                  NaN   
265      

Справочная инфомрация:
Нормальная частота дыхания у взрослых лошадей составляет от 8 до 12 вдохов в минуту, у новорожденных жеребят - от 60 до 80 вдохов в минуту. Жеребята в первые недели жизни сохраняют дыхательные нормы в пределах 20-40 вдохов в минуту.

Большая часть высоких значений данного столбца относится к молодым особям (до 6 месяцев), для которых характерно болеее частое дыхание.   
Также у особей с частым дыханием показатель частоты пульса тоже высокий. Что подтверждает правдивость этих данных.
Считаю данные выбросы укладываются в реальную картину и не требуют корректировки.


### Задание 3. Работа с пропусками
Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

Пропуски в столбцах с категориальными данными чаще всего заполняются модой.

In [79]:
for col in my_horse_colic_database_2.columns:
    pct_missing = my_horse_colic_database_2[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

_1_Surgery - 0.3%
_2_Age - 0.0%
_4_Temperature - 20.0%
_5_Pulse - 8.0%
_6_Respiratory_rate - 19.3%
_7_Temperature_of_extremities - 18.7%
_8_Peripheral_pulse - 23.0%
_9_Mucous_membranes - 15.7%


В столбце [1_Surgery] представлен категориальные данные: 1 - операция была, 2 - операции не было.
Пропущено всего 0,3% (одно значение). Будет логично заполнить его модой.

In [80]:
my_horse_colic_database_2['_1_Surgery'].value_counts()

1.0    180
2.0    119
Name: _1_Surgery, dtype: int64

In [81]:
my_horse_colic_database_na = my_horse_colic_database_2
my_horse_colic_database_2

_1_Surgery  _2_Age  _4_Temperature  _5_Pulse  _6_Respiratory_rate  \
0           2.0       1            38.5      66.0                 28.0   
1           1.0       1            39.2      88.0                 20.0   
2           2.0       1            38.3      40.0                 24.0   
3           1.0       2            39.1     164.0                 84.0   
4           2.0       1            37.3     104.0                 35.0   
..          ...     ...             ...       ...                  ...   
295         1.0       1             NaN     120.0                 70.0   
296         2.0       1            37.2      72.0                 24.0   
297         1.0       1            37.5      72.0                 30.0   
298         1.0       1            36.5     100.0                 24.0   
299         1.0       1            37.2      40.0                 20.0   

     _7_Temperature_of_extremities  _8_Peripheral_pulse  _9_Mucous_membranes  
0                              3.0                  3.0                  NaN  
1                              NaN                  NaN                  4.0  
2                              1.0                  1.0                  3.0  
3                              4.0                  1.0                  6.0  
4                              NaN                  NaN                  6.0  
..                             ...                  ...                  ...  
295                            4.0                  NaN                  4.0  
296                            3.0                  2.0                  4.0  
297                            4.0                  3.0                  4.0  
298                            3.0                  3.0                  3.0  
299                            NaN                  NaN                  NaN  

[300 rows x 8 columns]

In [103]:
my_horse_colic_database_na['_1_Surgery'].fillna(my_horse_colic_database_na['_1_Surgery'].mode()[0], inplace=True)
my_horse_colic_database_na

_1_Surgery  _2_Age  _4_Temperature  _5_Pulse  _6_Respiratory_rate  \
0           2.0       1           38.50      66.0                 28.0   
1           1.0       1           39.20      88.0                 20.0   
2           2.0       1           38.30      40.0                 24.0   
3           1.0       2           39.10     164.0                 84.0   
4           2.0       1           37.30     104.0                 35.0   
..          ...     ...             ...       ...                  ...   
295         1.0       1           38.35     120.0                 70.0   
296         2.0       1           37.20      72.0                 24.0   
297         1.0       1           37.50      72.0                 30.0   
298         1.0       1           36.50     100.0                 24.0   
299         1.0       1           37.20      40.0                 20.0   

     _7_Temperature_of_extremities  _8_Peripheral_pulse  _9_Mucous_membranes  
0                              3.0                  3.0                  1.0  
1                              NaN                  NaN                  4.0  
2                              1.0                  1.0                  3.0  
3                              4.0                  1.0                  6.0  
4                              NaN                  NaN                  6.0  
..                             ...                  ...                  ...  
295                            4.0                  NaN                  4.0  
296                            3.0                  2.0                  4.0  
297                            4.0                  3.0                  4.0  
298                            3.0                  3.0                  3.0  
299                            NaN                  NaN                  NaN  

[300 rows x 8 columns]

In [104]:
my_horse_colic_database_na['_1_Surgery'].value_counts()

1.0    181
2.0    119
Name: _1_Surgery, dtype: int64

#### В столбце [_2, Age] пропущенных данных нет

#### В столбце [_4, Temperature] температура пропущено 20% данных.
В описании к данным указано, что _7_Temperature_of_extremities корелирует с _4_Temperature. Заполним пропуски в _4_Temperature медианой в зависимости от значения _7_Temperature_of_extremities

In [105]:
my_horse_colic_database_na.loc[my_horse_colic_database_na['_4_Temperature'] == 38.5]


_1_Surgery  _2_Age  _4_Temperature  _5_Pulse  _6_Respiratory_rate  \
0           2.0       1            38.5      66.0                 28.0   
61          2.0       1            38.5      60.0                 24.0   
172         1.0       1            38.5     129.0                 48.0   
179         2.0       1            38.5     100.0                 24.0   
185         1.0       1            38.5     120.0                 60.0   
188         1.0       1            38.5      60.0                 24.0   
196         2.0       1            38.5      60.0                 20.0   
197         2.0       1            38.5      96.0                 36.0   
199         2.0       1            38.5      60.0                 40.0   
201         1.0       1            38.5      86.0                 24.0   
205         2.0       1            38.5     100.0                 24.0   
206         1.0       1            38.5      96.0                 30.0   
221         1.0       1            38.5      92.0                 40.0   
229         1.0       2            38.5     120.0                 70.0   
230         1.0       1            38.5     104.0                 40.0   
232         1.0       1            38.5      30.0                 18.0   
284         2.0       1            38.5      54.0                 24.0   
285         1.0       1            38.5      66.0                 24.0   
294         2.0       1            38.5      40.0                 16.0   

     _7_Temperature_of_extremities  _8_Peripheral_pulse  _9_Mucous_membranes  
0                              3.0                  3.0                  1.0  
61                             2.0                  2.0                  1.0  
172                            3.0                  3.0                  3.0  
179                            3.0                  3.0                  5.0  
185                            4.0                  3.0                  6.0  
188                            1.0                  1.0                  NaN  
196                            1.0                  1.0                  5.0  
197                            3.0                  3.0                  NaN  
199                            3.0                  1.0                  2.0  
201                            1.0                  1.0                  3.0  
205                            3.0                  3.0                  5.0  
206                            2.0                  3.0                  4.0  
221                            4.0                  3.0                  NaN  
229                            NaN                  NaN                  NaN  
230                            3.0                  3.0                  NaN  
232                            NaN                  NaN                  NaN  
284                            1.0                  1.0                  1.0  
285                            1.0                  1.0                  1.0  
294                            1.0                  1.0                  1.0

In [117]:
my_horse_colic_database_na4 = my_horse_colic_database_na
print(my_horse_colic_database_na.groupby('_7_Temperature_of_extremities')._4_Temperature.median())
my_horse_colic_database_na4._4_Temperature.fillna(my_horse_colic_database_na.groupby('_7_Temperature_of_extremities')._4_Temperature.transform('median'), inplace=True)

print(my_horse_colic_database_na4.groupby('_7_Temperature_of_extremities')._4_Temperature.median())

_7_Temperature_of_extremities
1.0    38.15
2.0    38.20
3.0    38.10
4.0    38.35
Name: _4_Temperature, dtype: float64
_7_Temperature_of_extremities
1.0    38.15
2.0    38.20
3.0    38.10
4.0    38.35
Name: _4_Temperature, dtype: float64


#### В столбце [_5_Pulse] пропущено 8 % значений. 
Также в нашем датафрейме есть столбец  [_8_Peripheral_pulse] (Периферический пульс). Будет логично предположить что есть некоторая взаимная зависимость данных параметров. Заполним пропуски в рассматриваемом столбце медианой для каждого из  
[_8_Peripheral_pulse]

In [119]:
my_horse_colic_database_na4
print(my_horse_colic_database_na.groupby('_8_Peripheral_pulse')._5_Pulse.median())
my_horse_colic_database_na4._5_Pulse.fillna(my_horse_colic_database_na.groupby('_8_Peripheral_pulse')._5_Pulse.transform('median'), inplace=True)

_8_Peripheral_pulse
1.0    52.0
2.0    60.0
3.0    88.0
4.0    88.0
Name: _5_Pulse, dtype: float64


#### В столбце [_6_Respiratory_rate ] пропущено 19.3% значений
Изучив предметную область выяснилось, что частота дыхания зависит от возраста. Также заполним медианой для каждого возраста. 


In [109]:
my_horse_colic_database_na4
print(my_horse_colic_database_na.groupby('_2_Age')._6_Respiratory_rate.median())
my_horse_colic_database_na4._6_Respiratory_rate.fillna(my_horse_colic_database_na.groupby('_2_Age')._6_Respiratory_rate.transform('median'), inplace=True)

_2_Age
1    24.0
2    49.0
Name: _6_Respiratory_rate, dtype: float64


#### В столбце [_7_Temperature_of_extremities ] пропущено 18.7% значений
Это категориальный параметр. Заполняем модой.


In [114]:
my_horse_colic_database_na4['_7_Temperature_of_extremities'].fillna(my_horse_colic_database_na4['_7_Temperature_of_extremities'].mode()[0], inplace=True)


#### В столбце [_8_Peripheral_pulse] пропущено 23% значений
Это категориальный параметр. Заполняем модой.

In [112]:
my_horse_colic_database_na4['_8_Peripheral_pulse'].fillna(my_horse_colic_database_na4['_8_Peripheral_pulse'].mode()[0], inplace=True)

#### В столбце [_9_Mucous_membranes] пропущено 15,7% значений
Это категориальный параметр. Заполняем модой.

In [113]:
my_horse_colic_database_na4['_9_Mucous_membranes'].fillna(my_horse_colic_database_na4['_9_Mucous_membranes'].mode()[0], inplace=True)

In [115]:
my_horse_colic_database_na4

_1_Surgery  _2_Age  _4_Temperature  _5_Pulse  _6_Respiratory_rate  \
0           2.0       1           38.50      66.0                 28.0   
1           1.0       1           39.20      88.0                 20.0   
2           2.0       1           38.30      40.0                 24.0   
3           1.0       2           39.10     164.0                 84.0   
4           2.0       1           37.30     104.0                 35.0   
..          ...     ...             ...       ...                  ...   
295         1.0       1           38.35     120.0                 70.0   
296         2.0       1           37.20      72.0                 24.0   
297         1.0       1           37.50      72.0                 30.0   
298         1.0       1           36.50     100.0                 24.0   
299         1.0       1           37.20      40.0                 20.0   

     _7_Temperature_of_extremities  _8_Peripheral_pulse  _9_Mucous_membranes  
0                              3.0                  3.0                  1.0  
1                              3.0                  1.0                  4.0  
2                              1.0                  1.0                  3.0  
3                              4.0                  1.0                  6.0  
4                              3.0                  1.0                  6.0  
..                             ...                  ...                  ...  
295                            4.0                  1.0                  4.0  
296                            3.0                  2.0                  4.0  
297                            4.0                  3.0                  4.0  
298                            3.0                  3.0                  3.0  
299                            3.0                  1.0                  1.0  

[300 rows x 8 columns]

In [120]:
for col in my_horse_colic_database_na4.columns:
    pct_missing = my_horse_colic_database_na4[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

_1_Surgery - 0.0%
_2_Age - 0.0%
_4_Temperature - 0.0%
_5_Pulse - 0.0%
_6_Respiratory_rate - 0.0%
_7_Temperature_of_extremities - 0.0%
_8_Peripheral_pulse - 0.0%
_9_Mucous_membranes - 0.0%
